### ISSUE TITLE
Unify Adaptive elements' delete / create / copy behavior in Composer / Form / Visual

### ISSUE THREAD

**Issue by yeze322** [microsoft/BotFramework-Composer#2312]
## Is your feature request related to a problem? Please describe.

In Visual Editor, delete / create / copy behavior are hooked with LU / LG resources heavily;
in Form Editor & composer, if we want a unified behavior, we have to implement the same thing again by copying the code.

Now that we are hooking dialog operations with more external resources such as LG / LU or even QnA, we really need a shared lib to handle the complexity of dialog element changes.

## Describe the solution you'd like

Have a shared dialog CRUD lib and hook the CRUD handlers with LG / LU resource management API.
Share the lib across Shell / Visual / Form

**Comment by yeze322** [comment]
refs #2273 #2209

**Comment by yeze322** [comment]
### Adaptive dialog CRUD operations checklist

#### Visual Editor
1. **DELETE** **_action_** via '...' flyout
2. **ADD** **_action_** via '+' flyout
3. **PASTE** **_action_** via '+' flyout

#### Form Editor
1. **UPDATE** **_action_**
2. **UPDATE** **_trigger_**
3. **UPDATE** **_dialog_**

#### Shell - Toolbar
1. **COPY** **_action_** selection
2. **CUT** **_action_** selection
3. **MOVE** **_action_** selection + **ADD** target **_trigger_** via modal
4. **DELETE** **_action_** selection

#### Shell - ProjectTree
1. **ADD** **_trigger_** via 'New Trigger...' modal
2. **ADD** **_dialog_** via 'New Dialog...' modal
3. **DELETE** **_trigger_** via '...' flyout
4. **DELETE** **_dialog_** via '...' flyout

### Resource types X Operations
1. **_Dialog_** x **Add** | **Update** | **Delete**
2. **_Trigger_** x **Add** | **Update** | **Delete**
3. **_Action_** x **Add** | **Update** | **Delete** | **Copy** | **Paste** | **Move**

@lei9444 @a-b-r-o-w-n @cwhitten please review, hope I didn't miss any operation

**Comment by boydc2014** [comment]
@yeze322  the list looks good, one question is i think we were talking about if it's possible to be done entirely in Shell, totally transparent to Visual and Form.

Is that option still valid or not working so that we need a shared lib called by different component?

**Comment by yeze322** [comment]
@boydc2014 This item is important but not necessarily to be included in R9.
I tagged it with 'R9' weeks ago, I also prefer to postpone it to R10

**Comment by boydc2014** [comment]
> @boydc2014 This item is important but not necessarily to be included in R9.
> I tagged it with 'R9' weeks ago, I also prefer to postpone it to R10
> 
> It's a progressive work.

I think it's OK to do it systematically in R10, the thing is for LG, it's actually ok, even if we don't delete some content, but for LU, it's better to clean up when deleting input and trigger, because anything left in .lu file will affect model's behavior because it would be trained together. I will have Leilei to fix the LU behaviors first. Make sense?

**Comment by boydc2014** [comment]
Also, talking about a good solution in R10, I will vote for a transparent layer (like some hooks or listeners of dialog json updates) maybe combine some "virtual property" for LU and LG, if possible. A shared lib is less ideal because it still explicitly managing "side effects", which seems like a leak of abstraction.  What do you think? Is there any operations must be exposed to editor directly?

**Comment by a-b-r-o-w-n** [comment]
I think that the extension package would be great for exposing these apis. It's the connection between the shell and the extension and has a clear boundary. It already exposes a `useShellApi` hook, but it would be nice to have a set of hooks like `useDialogApi` `useActionApi` etc that can apply additional layers of performance optimization (memoization, etc). Basically, take the shell's state and actions and divide them by domain and memoize.

**Comment by zhixzhan** [comment]
I think we can centralize incoming CRUD request at dialog level (dialog driven), looks like
<img width="412" alt="image" src="https://user-images.githubusercontent.com/49866537/79323361-ade01600-7f40-11ea-9f76-f2bee52e1b4f.png">

When actions.dialog accept a request
add:
```
{
     "$kind": "Microsoft.SendActivity",
     "$designer": {
            "id": "tzI6jE"
      },
      "activity": "${bfdactivity-tzI6jE()}"
}
```
dialog action center would trigger an `add lg action`, same to delete/copy.

**Comment by yeze322** [comment]
Considerations:
1. **Abstraction**: Managing side effects outside editor @boydc2014 
    - motivation -  avoid leak of abstraction
    - currently - all action operations are managed by Visual Editor
    - todo - we need improvements to not expose operations to editor directly
2. **Plugin**: Extension packages should handle additional logic @a-b-r-o-w-n 
    - motivation - plugin system
    - currently - LG resource is manually maintained in Composer-core
    - todo - under the plugin pattern, it's 'ui-plugins/lg' that provide LG controllers (delete/copy/create)
3. **CRUD API**: Have a centralized dialog CRUD layer shared by Form/Visual/Shell @zhixzhan @ Benjamin who works on ScreenReader
    - motivation - unified CRUD experience
    - currently - actions managed by Visual, triggers & dialogs managed by Shell
    - todo - Shell should provide basic json operations shared by Shell / Visual / Form, then we can hook after an operation to notify downstream consumer such as
      - notify Screen Reader to read something
      - if special resources LG / LU are detected to be changed, trigger extra CRUD APIs
4. **Customized action**: Schema-driven copy & paste & delete @yeze322 
    - motivation - handle LG / LU resources in cutomized action automatically
    - currently - constructors & destructors are manually maintained, if a customized action contains `IActivityTemplate`, it cannot be handled automatically
    - todo - copy & paste & delete should be schema driven, hooked after the centralized dialog center, if an action is deleted, a schema checker will try invoking the registered destructor that matches given field type


In R9, we can focus on the **Plugin** part (it answers **Abstraction**), figure out what should be provided by a plugin (e.g. LG / LU) to serve the Aspect of dialog operation.

**Comment by yeze322** [comment]
I suggest we think it in AOP but start from a very basic context + callback pattern
```ts
const copyAction = (dialog, actionId) => {
  const { copyConstructor } = useContext(ActionApi);
  const action = get(dialog, actionId);
  const copy = copyConstructor(action)
  return copy;
}

const copyConstructor = (action) => {
  // @ before
  // copy
  
  // @ after
}
```

**Comment by boydc2014** [comment]
Thank @yeze322 for thoughtful considerations, and to @zhixzhan let dialog directs the operations is good. 

From a high level, there are two principles i'm thinking it would be great if we can hold

### 1  Make editors care as less as possible, call APIs as less as possible

In that sense, LG and LU related update should be managed totally as `side effect` and totally transparent to editors, if possible. I think most cases, like create a node, are possible and also very straightforward, i will address how it can be supported in our hoisted experiences. See later section for "Virtual Property".

From an editor's POV, the editor calls dialog api, create, update, copy, delete, whatever operations on whatever node (trigger or action),  all the related side effects should be happen behind the scene. 

Regardless how this is done, eventually the criteria for checking if we meet this principle, is there is no lg\lu related api call on normal operations of those editors. 

### 2 Manage side effects through a clean and decoupled pattern

Then we move to how to manage side effects, I think we need a better pattern for this, instead of manually trigger lg\lu actions again inside dialog api, which is still not scalable. 

A pub-sub pattern through an event system seems perfect fine for this case. The flow i can image is 
* Editor calls dialogApi.createStep(initial payload)
* the API is invoked (shellAPI or dialogAPI) dialog content is updated, and a event is generated
* there are two subscribers to this event, lg, lu, which will all process this event and do proper processing. screen reader can also be one of them. 

> we probably can even also include dialog as subscriber, then our api action becomes a pure event generator, our solution is more like a pure `event sourcing` system. 

#### Virtual Property for hoisted contents

[To be filled]

**Comment by yeze322** [comment]
@boydc2014 Inspired by AOP, we can also consider introducing decorators in TS.

Pub-sub model can solve side effect problems, but in some cases, we need middlewares or `@before` processor + `@after` processor.

As an example, when copying a SendActivity, current logic is:
```ts
const copySendActivity = (action) => {
  const copy = deepClone(action);
  const lgContent = getLgTemplate(copy.activity);
  copy.activity = lgContent;
  return activity;
}
```
To decouple LG effects, it could be
```ts
@after('replaceLgToContent')
@after('readScreen')
function copySendActivity (action) {
  return deepClone(action);
}
```
The problem in a stream pattern is, plugins have no reason to subscribe the event, so the complexity is still in Composer. What a plugin can provide may be middlewares before / after COPY | DELETE | ...
```ts
const LGPlugin = {
  middlewares: {
    copy: {
      after: (action) => replaceLgToContent(action)
    }
  }
}

const mergePlugins = (plugins) => {
  const beforeCopyHandlers = plugins.map(x => x.middlewares.copy.before);
  const beforeCopyHandlers = plugins.map(x => x.middlewares.copy.after);
  const copy = action => {
    const action' = before(action);
    ActionApi.copy(action');
    return after(action');
  }
}
```

**Comment by yeze322** [comment]
@boydc2014 yes, my high priority in R9 is just fixing the LU resource behavior, I will use a traditional callback pattern to achieve that

**Comment by boydc2014** [comment]
I did't get why a Pub-Sub system won't work in your case of copy LgActions, i would image 

```
function copySendActivity(action: SendAcitvity)
{ 
  // step 1: clone the json
  const clone = deepClone(action);
  // step 2:  regen id
  clone.designer.id =genid() 
  // step 3: copy is just create
  dialogAPI.addNode(clone, path)
}

function addNode(node, path)
{
 emitEvent( {type: NodeAdded, path: path})
}
```

I think this should work  if we take care of the rest of lg\lu. 

Also, i don't know why "it has no reason that plugin subscribe to events", let plugin subsribe to events is something very common, like vscode, all the extensions in vscode has certain capabilities to subscribe to a wide range of events. 

If i understand correctly, you are refering to that the editor (also a plugin) itself have to know when certain action is finished, right? 

If some thing has to wait until all side effects to be finished, we can emit a DONE event and the sender (who init the action) can get notified.  But i really doubt that, we will have many or even some cases that we have to wait the side effects to be totally done. 


Speaking of AOP and decorators, i think it looks neat by looking the code it delivered as samples. But based on my experience on C# and python decorators, i don't think it should heavily used for normal coding tasks. My understanding of AOP, it focus on solving an `aspect`, which usually each aspect is a totally different view of the product, like you will treat "logging" or "debugging" as an aspect. But hardly we want to treat the pre\after of an action, inside a normal coding path as an "aspect".  I assume probably when you say inspired by AOP, maybe you are only referring the decorator patterns, I will love to give a try see how that can simplify code, if we are only talking about coding patterns. 

Speaking of Pub-Sub and event system, i think that's something proven to deliver good decoupling that helps separate the concern. 

In the high level of this design issue, at the end the day, the question i would ask is not we use pattern A vs pattern B, is that
* Are we making each component more focused
   * by reducing the number of components that it talks to or
   * by simplifying the way how a component talk to the other component

**Comment by cwhitten** [comment]
https://github.com/microsoft/BotFramework-Composer/pull/2718
